### PR TITLE
API refactor (memory, offsets, wg counts)

### PR DESCRIPTION
**Pull request by kvark**
Improved API consistency.
PR checklist:
- [x] `make` succeeds (on *nix)
- [x] `make reftests` succeeds
- [x] tested examples with the following backends: Vulkan,GL

### CODE REVIEW

File: src/backend/vulkan/src/command.rs [hunk 1]
```diff
@@ -224,8 +241,8 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
                 self.raw, // commandBuffer
                 conv::map_pipeline_stage(stages.start),
                 conv::map_pipeline_stage(stages.end),
-                vk::DependencyFlags::empty(), // dependencyFlags // TODO
-                &[],
+                mem::transmute(dependencies),
```
Comment (grovesNL) [inline src/backend/vulkan/src/command.rs]: Could we use `dependencies.bits()` or does it not exist here?

Comment (kvark) [reply src/backend/vulkan/src/command.rs]: It expects a bit-like opaque [structure](https://docs.rs/ash/0.21.1/ash/vk/types/struct.DependencyFlags.html) here that doesn't have `from_bits` (yet?). I'd honestly prefer ash just use the stock `bitflags` crate.

Comment (grovesNL) [reply src/backend/vulkan/src/command.rs]: Oh ok, makes sense

File: src/backend/dx12/src/command.rs [hunk 2]
```diff
@@ -705,8 +705,19 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
...
-            match *barrier {
+            match *barrier.borrow() {
+                memory::Barrier::AllBuffers(_) |
+                memory::Barrier::AllImages(_) => {
+                    let mut bar = d3d12::D3D12_RESOURCE_BARRIER {
+                        Type: d3d12::D3D12_RESOURCE_BARRIER_TYPE_UAV,
+                        Flags: d3d12::D3D12_RESOURCE_BARRIER_FLAG_NONE,
+                        u: unsafe { mem::zeroed() },
+                    };
+                    *unsafe { bar.u.UAV_mut() } = d3d12::D3D12_RESOURCE_UAV_BARRIER {
```
Comment (msiglreith) [inline src/backend/dx12/src/command.rs]: So we have confirmation that it's fine with before-after state mismatches?

Comment (msiglreith) [reply src/backend/dx12/src/command.rs]: Could you also add a comment here that UAV barriers provide the necessary sync behaviour?

File: src/hal/src/buffer.rs [hunk 3]
```diff
@@ -5,6 +5,10 @@ use std::fmt;
 
 use {IndexType, Backend};
 
+
+/// An offset inside a vertex buffer, in bytes.
```
Comment (msiglreith) [inline src/hal/src/buffer.rs]: Why vertex buffer only?

File: src/hal/src/memory.rs [hunk 4]
```diff
@@ -62,9 +62,24 @@ bitflags!(
...
 );
 
+bitflags!(
+    /// Barrier dependency flags.
+    #[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+    pub struct Dependencies: u32 {
+        ///
+        const BY_REGION    = 0x1;
+        ///
+        const VIEW_LOCAL   = 0x2;
```
Comment (msiglreith) [inline src/hal/src/memory.rs]: Should we indicate that these are part of extensions?

**Comment by kvark** [comment]
(glutin breaking changes...)
bors r=msiglreith

**Comment by kvark** [comment]
bors r=msiglreith

**Pull request merged**